src/librustdoc/doctest/make.rs

```diff
@@ -184,6 +184,64 @@ pub(crate) struct DocTestBuilder {
     pub(crate) can_be_merged: bool,
 }
 
+/// Contains needed information for doctest to be correctly generated with expected "wrapping".
+pub(crate) struct WrapperInfo {
+    pub(crate) before: String,
+    pub(crate) after: String,
+    pub(crate) returns_result: bool,
+    insert_indent_space: bool,
+}
+
+impl WrapperInfo {
+    fn len(&self) -> usize {
+        self.before.len() + self.after.len()
+    }
+}
+
+/// Contains a doctest information. Can be converted into code with the `to_string()` method.
+pub(crate) enum DocTestWrapper {
+    Valid { crate_level_code: String, wrapper: Option<WrapperInfo>, code: String },
+    SyntaxError(String),
+}
+
+impl std::string::ToString for DocTestWrapper {
+    fn to_string(&self) -> String {
+        match self {
+            Self::SyntaxError(s) => s.clone(),
+            Self::Valid { crate_level_code, wrapper, code } => {
+                let prog_len = code.len()
+                    + crate_level_code.len()
+                    + wrapper.as_ref().map(|w| w.len()).unwrap_or(0);
```

Comment:
If we're doing all this to estimate capacity, should we also do `+ 4 * code.lines().count()`, or is that counterproductive? how much do we care about avoiding allocations?

Reply:
We have the whole content, so it'd be a shame not to use it. However I missed the case where we want to indent indeed.